cupyx/scipy/spatial/_kdtree_utils.py

```diff
@@ -0,0 +1,929 @@
+        const double* __restrict__ box_bounds,
+        const double* __restrict__ tree_bounds,
+        long long* all_nodes, long long* node_count,
+        long long* all_debug_nodes) {
+
+    const int idx = blockIdx.x * blockDim.x + threadIdx.x;
+    if(idx >= points_size) {
+        return;
+    }
+
+    double* point = points + n_dims * idx;
+    long long* nodes = all_nodes + n * idx;
+    long long* debug_nodes = all_debug_nodes + 3 * n * idx;
+
+    adjust_to_box(point, n_dims, box_bounds);
+    long long count = compute_query_ball<double>(
+        n, n_dims, radius, eps, p, true, sort, point, tree, index, box_bounds,
+        tree_bounds, nodes, debug_nodes);
+
+    node_count[idx] = count;
+}
+'''
+
+
+KD_MODULE = cupy.RawModule(
+    code=KD_KERNEL, options=('-std=c++11',),
+    name_expressions=['update_tags', 'tag_pairs'] + [
+        f'compute_bounds<{x}>' for x in TYPE_NAMES])
+
+KNN_MODULE = cupy.RawModule(
+    code=KNN_KERNEL, options=('-std=c++11',),
+    name_expressions=['knn_periodic', 'query_ball_periodic'] +
+    [f'knn<{x}>' for x in TYPE_NAMES] +
+    [f'query_ball<{x}>' for x in TYPE_NAMES])
+
+
+def _get_module_func(module, func_name, *template_args):
+    args_dtypes = [_get_typename(arg.dtype) for arg in template_args]
+    template = ', '.join(args_dtypes)
+    kernel_name = f'{func_name}<{template}>' if template_args else func_name
+    kernel = module.get_function(kernel_name)
+    return kernel
+
+
+def asm_kd_tree(points):
+    """
+    Build an array-based KD-Tree from a given set of points.
+
+    Parameters
+    ----------
+    points: ndarray
+        Array input of size (m, n) which contains `m` points with dimension
+        `n`.
+
+    Returns
+    -------
+    tree: ndarray
+        An array representation of a left balanced, dimension alternating
+        KD-Tree of the input points.
+    indices: ndarray
+        An index array that maps the original input to its corresponding
+        KD-Tree representation.
+
+    Notes
+    -----
+    This algorithm is derived from [1]_.
+
+    References
+    ----------
+    .. [1] Wald, I., GPU-friendly, Parallel, and (Almost-)In-Place
+           Construction of Left-Balanced k-d Trees, 2022.
+           doi:10.48550/arXiv.2211.00120.
+    """
+    x = points.copy()
+    track_idx = cupy.arange(x.shape[0])
+    tags = cupy.zeros(x.shape[0], dtype=cupy.int64)
+    length = x.shape[0]
+    dims = x.shape[1]
+    n_iter = int(np.log2(length))
+
+    block_sz = 128
+    n_blocks = (length + block_sz - 1) // block_sz
+    update_tags = KD_MODULE.get_function('update_tags')
+    x_tags = cupy.empty((2, length), dtype=x.dtype)
+
+    level = 0
+    for level in range(n_iter):
+        dim = level % dims
+        x_tags[0, :] = x[:, dim]
+        x_tags[1, :] = tags
+        idx = cupy.lexsort(x_tags)
+        x = x[idx]
+        tags = tags[idx]
+        track_idx = track_idx[idx]
+        update_tags((n_blocks,), (block_sz,), (length, level, tags))
```

Comment:
I can find `thrust::sort` declarations with `__device__` specified. I wonder if it could be called in `RawKernel`, but I have never tried that for myself. Would you take a look at these?
https://thrust.github.io/doc/group__sorting_ga3f47925d80f4970d5730051dba1c5603.html
https://thrust.github.io/doc/group__sorting_gaef9372b75a6920c67c43150627e74225.html

Reply:
If it doesn't work, let us keep using `lexsort` here. We wouldn't like to have dynamic parallelism this part.

Reply:
> I can find `thrust::sort` declarations with `__device__` specified. I wonder if it could be called in `RawKernel`, but I have never tried that for myself. 

According to some sources in other forums, it seems that a device kernel can call `thrust::sort` under the following conditions:

1. It serializes the execution to that single kernel, by setting `thrust::seq` as policy → No parallel gains in our case
2. It spawns further kernels via dynamic parallelism, by setting `thrust::device` as policy → Parallel gains, but dynamic parallelism is involved.
  
In this case our ideal solution would be to define a kernel that performs the sorting process in a single pass with a custom comparator instead of two, which is the current `lexsort` behaviour. I don't know if we can define a pair sorting function that calls `thrust::sort_by_key`, similar to https://github.com/cupy/cupy/blob/1cd6db5566eab966f7d4cd96bf701dba81e784c5/cupy/cuda/cupy_thrust.cu#L388 in a Cython file or similar

Reply:
Thanks for the investigation. Let me keep the current implementation using `lexsort` for this moment, and revisit in another issue.